conftest.py

```diff
@@ -51,3 +55,47 @@ def pytest_itemcollected(item):
 
     if isinstance(item, pytest.DoctestItem):
         item.add_marker(doctest_marker)
+
+
+def pytest_addoption(parser: pytest.Parser) -> None:
+    """Hook to register custom CLI options and config values"""
+    parser.addoption(
+        "--strict-warnings",
+        action="store_true",
+        default=False,
+        help="Turn most warnings into errors",
+    )
+
+    parser.addini(
+        "filterwarnings_when_strict",
+        "Filters to apply after `-Werror` when --strict-warnings is active",
+        type="linelist",
+        default=[],
+    )
+
+
+def pytest_collection_modifyitems(
+    session: pytest.Session, config: pytest.Config, items: Iterable[pytest.Item]
+):
+    if not config.getoption("--strict-warnings"):
+        return
+
+    warning_filters = [
+        r"error",
+        *_config_get_strlist(config, "filterwarnings"),
+        *_config_get_strlist(config, "filterwarnings_when_strict"),
+    ]
+    warning_marks = [pytest.mark.filterwarnings(f) for f in warning_filters]
+
+    for item in items:
+        # reverse and then prepend means essentially `marks[0:0] = warning_marks`
+        # this ensures that markers that are applied later override these
+        for mark in reversed(warning_marks):
+            item.add_marker(mark, append=False)
+
+
+def _config_get_strlist(config: pytest.Config, name: str) -> list[str]:
+    if strs := config.getini(name):
+        assert isinstance(strs, list)
+        return [str(f) for f in strs]
```

Comment:
Is `str(f)` necessary?  These could somehow not be strings?  If so, maybe the list shouldn't be called `strs`?